service/worker/archiver/activities.go

```diff
@@ -225,6 +232,53 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
 		logger.Error(uploadErrorMsg, tag.ArchivalUploadFailReason(errorDetails(err)), tag.ArchivalBlobKey(indexBlobKey.String()), tag.Error(err))
 		return err
 	}
+	if runBlobIntegrityCheck {
+		scope.IncCounter(metrics.ArchiverRunningBlobIntegrityCheckCount)
+		req := &clientShared.GetWorkflowExecutionHistoryRequest{
+			Domain: common.StringPtr(request.DomainName),
+			Execution: &clientShared.WorkflowExecution{
+				WorkflowId: common.StringPtr(request.WorkflowID),
+				RunId:      common.StringPtr(request.RunID),
+			},
+		}
+		resp, err := container.PublicClient.GetWorkflowExecutionHistory(ctx, req)
+		if err != nil {
+			scope.IncCounter(metrics.ArchiverCouldNotRunBlobIntegrityCheckCount)
+			logger.Error("failed to access history for blob integrity check", tag.Error(err))
+			return nil
+		}
+		if !resp.GetArchived() {
+			scope.IncCounter(metrics.ArchiverBlobIntegrityCheckFailedCount)
+			logger.Error("blob integrity check failed history was not archived")
+			return nil
+		}
+		fetchedHistory := resp.History
+		req.NextPageToken = resp.NextPageToken
+		for len(req.NextPageToken) != 0 {
+			resp, err := container.PublicClient.GetWorkflowExecutionHistory(ctx, req)
+			if err != nil {
+				scope.IncCounter(metrics.ArchiverCouldNotRunBlobIntegrityCheckCount)
+				logger.Error("failed to access history for blob integrity check", tag.Error(err))
+				return nil
+			}
+			if !resp.GetArchived() {
+				scope.IncCounter(metrics.ArchiverBlobIntegrityCheckFailedCount)
+				logger.Error("blob integrity check failed history was not archived")
+				return nil
+			}
+			fetchedHistory.Events = append(fetchedHistory.Events, resp.History.Events...)
+			req.NextPageToken = resp.NextPageToken
+		}
+		equal, reason := historiesEqual(fetchedHistory, uploadedHistory)
+		if err != nil {
```

Comment:
this can never happen, you already return from the method on all error paths above. This code can be removed